libs/wire-api-federation/src/Wire/API/Federation/API/Galley.hs

```diff
@@ -19,40 +19,63 @@
+import Data.Qualified (Qualified)
+import Imports
 import Servant.API (JSON, Post, ReqBody, (:>))
 import Servant.API.Generic ((:-))
+import Wire.API.Arbitrary (Arbitrary, GenericUniform (..))
+import Wire.API.Conversation (Conversation)
+import Wire.API.Federation.Util.Aeson (CustomEncoded (CustomEncoded))
+
+-- FUTUREWORK: data types, json instances, more endpoints. See
+-- https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/356090113/Federation+Galley+Conversation+API
+-- for the current list we need.
 
 data Api routes = Api
-  { getConversation ::
+  { getConversations ::
       routes
         :- "federation"
         :> "conversations"
-        -- usecases:
-        -- - e.g. upon registering a new client to your account, get the list of your conversations
-        :> "list-conversations"
-        :> ReqBody '[JSON] ListConversations
-        :> Post '[JSON] ListConversationsResponse,
-    conversationMemberChange ::
+        :> "get-by-ids"
+        :> ReqBody '[JSON] GetConversationsRequest
+        :> Post '[JSON] [Conversation],
+    -- used by backend that owns the conversation
+    -- to inform federating backends about changes
+    updateConversationMembership ::
       routes
         :- "federation"
         :> "conversations"
-        -- for the usecase:
-        -- given alice,alice2 @A and bob @B:
-        -- alice adds Bob: /add-to-conversation(bob)@A
-        --   A -> B: check bob exists on B
-        --   A: add B to conversation database entry
-        --   A -> B: by the way, B is now in one of my conversations.
-        --   (B writes this in its DB: "Bob exists in a conversation with ID 1 in A)
-        :> "member-change"
-        :> ReqBody '[JSON] ConversationMemberChange
-        :> Post '[JSON] ConversationMemberChangeResponse
+        :> "update-membership"
+        :> ReqBody '[JSON] ConversationMemberUpdate
+        :> Post '[JSON] ()
+  }
+  deriving (Generic)
+
+data GetConversationsRequest = GetConversationsRequest
+  { gcrUserId :: UserId,
+    gcrConvIds :: [ConvId]
   }
+  deriving stock (Eq, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform GetConversationsRequest)
+  deriving (ToJSON, FromJSON) via (CustomEncoded GetConversationsRequest)
 
--- FUTUREWORK: data types, json instances, more endpoints. See https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/356090113/Federation+Galley+Conversation+API for the current list we need.
-type ConversationMemberChange = ()
+data MembershipUpdateOp
+  = RemoveMember
+  | AddMember
+  deriving stock (Eq, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform MembershipUpdateOp)
 
-type ConversationMemberChangeResponse = ()
+instance ToJSON MembershipUpdateOp
 
-type ListConversations = ()
+instance FromJSON MembershipUpdateOp
 
-type ListConversationsResponse = ()
+data ConversationMemberUpdate = ConversationMemberUpdate
+  { cmuConvId :: Qualified ConvId,
+    cmuUserId :: UserId,
+    cmuOperation :: MembershipUpdateOp
```

Comment:
We need support for adding multiple users in the same request. Perhaps we could have a list of users to remove and a list of users to add?